Fix: get_format accepts lazy format strings (date filter regression) [swev-id: django__django-15741]

casey-brooks opened the pull request:
This PR fixes a regression where lazy format strings passed via the date template filter cause:

`TypeError: getattr(): attribute name must be string` in `django.utils.formats.get_format`.

- Implements coercion of non-`str` inputs (lazy gettext `Promise`, `bytes`) to native `str` via `force_str` at the top of `get_format`.
- Adds tests covering template date filter with a lazy format and direct `get_format` cases (lazy, bytes, non-ASCII, SafeString).

Linked Issue: #547

Reproduction steps:
```python
import datetime
from django.conf import settings
settings.configure(USE_L10N=True, USE_I18N=True, LANGUAGE_CODE='en', INSTALLED_APPS=[], TIME_ZONE='UTC')
import django; django.setup()
from django.utils.translation import gettext_lazy as _
from django.utils import formats
d = datetime.date(2020,1,1)
formats.date_format(d, _('Y-m-d'))
```

Observed stack trace excerpt:
```
TypeError: getattr(): attribute name must be string
  File "django/utils/formats.py", line 158, in date_format
    value, get_format(format or "DATE_FORMAT", use_l10n=use_l10n)
  File "django/utils/formats.py", line 128, in get_format
    val = getattr(module, format_type, None)
```

Notes:
- Base branch: django__django-15741
- CI is intentionally not run per project instructions.

casey-brooks committed:
fix(formats): coerce get_format inputs

Handle lazy, bytes, and SafeString format identifiers via force_str at the top of get_format and cover regressions with template and utility tests.

Fixes #547.

casey-brooks commented:
## Test & Lint Summary
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py tests.template_tests.filter_tests.test_date --parallel=1` → 14 passed
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py tests.utils_tests.test_formats_get_format --parallel=1` → 5 passed
- `flake8 django/utils/formats.py tests/template_tests/filter_tests/test_date.py tests/utils_tests/test_formats_get_format.py` → no issues

noa-lucent commented on django/utils/formats.py:
```diff
@@ -106,6 +107,8 @@ def get_format(format_type, lang=None, use_l10n=None):
     """
+    if not isinstance(format_type, str):
+        format_type = force_str(format_type)
     if use_l10n is None:
```
[major] Calling `force_str()` without `strings_only=True` changes longstanding behavior for callers that pass protected types like `None`, numbers, or dates. Before this patch, `get_format(None)`, `get_format(10)`, etc. simply returned the original object because they fall through to the `return format_type` branch. After the conversion, those inputs become the string literal `'None'`, `'10'`, etc., which will also poison the cache key. Even if the public contract expects a string name, the function has historically tolerated these cases and we should avoid regressing them. Could you switch this to `force_str(format_type, strings_only=True)` and add a quick regression test (e.g., asserting `get_format(None)` is still `None`) so we don’t lose this behavior again?

noa-lucent reviewed (changes requested):
Thanks for addressing the regression and adding coverage. I left one blocking note: we need to preserve get_format's historic behavior for protected non-string inputs (e.g., None) by using force_str(..., strings_only=True) and covering that case. Once that's in place I'm happy to take another look.

casey-brooks committed:
fix(formats): preserve protected get_format inputs

Use force_str(strings_only=True) to avoid coercing protected types like None and add regression coverage to ensure the old behavior remains.

casey-brooks commented:
## Test & Lint Summary (update)
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py tests.template_tests.filter_tests.test_date --parallel=1` → 14 passed
- `PYTHONPATH=$PWD DJANGO_SETTINGS_MODULE=tests.test_sqlite python tests/runtests.py tests.utils_tests.test_formats_get_format --parallel=1` → 6 passed
- `flake8 django/utils/formats.py tests/template_tests/filter_tests/test_date.py tests/utils_tests/test_formats_get_format.py` → no issues

noa-lucent reviewed (approved):
Thanks for tightening the coercion and adding the regression test—everything looks good to me now.
